otter/convergence/service.py

```diff
@@ -799,3 +808,18 @@ def divergent_changed(self, children):
             # the return value is ignored, but we return this for testing
             eff = self._converge_all(my_buckets, children)
             return perform(self._dispatcher, self._with_conv_runid(eff))
+
+
+class ConvergenceExecutor(object):
+    gather = attr.ib()
+    plan = attr.ib()
+    get_desired_group_state = attr.ib()
+    convergence_succeeded = attr.ib()
```

Comment:
I think the thing here is that `convergence_failed` doesn't actually have any logic that's specific to either launch_server or heat stacks, whereas `convergence_succeeded` does -- by updating the servers cache.

I'm a little bit concerned about this factoring, and think there may be a better way to do it. For example, it looks like the *majority* of the code in `convergence_succeeded_servers` will need to be used in the heat case too, just not the servers-cache updating. So maybe it should be refactored so that `convergence_succeeded` is just a fixed function that's called (not dispatched in this `ConvergenceExecutor` object), and then maybe we could have an `after_success` method that, for launch_server, will update the cache and for heat will do nothing? (or maybe update the cache in a different way!)